transform/coroutines.go

```diff
@@ -0,0 +1,934 @@
+			}
+		case llvm.Unreachable:
+			prev := llvm.PrevInstruction(last)
+
+			if prev.IsNil() || prev.IsACallInst().IsNil() || !c.isAsyncCall(prev) {
+				// This unreachable instruction does not behave as an asynchronous return.
+				continue
+			}
+
+			// This is an asyncnhronous tail call to function that does not return.
+			returns[bb] = returnDeadTail
+		}
+	}
+
+	fn.returns = returns
+}
+
+// returnAnalysisPass runs an analysis pass which classifies the returns of all async functions.
+func (c *coroutineLoweringPass) returnAnalysisPass() {
+	for _, async := range c.asyncFuncs {
+		c.analyzeFuncReturns(async)
+	}
+}
+
+// categorizeCalls categorizes all asynchronous calls into regular vs. async and matches them to their callers.
+func (c *coroutineLoweringPass) categorizeCalls() {
+	// Sort calls into their respective callers.
+	for _, async := range c.asyncFuncs {
+		async.calls = map[llvm.Value]struct{}{}
+	}
+	for _, call := range c.calls {
+		c.asyncFuncs[call.InstructionParent().Parent()].calls[call] = struct{}{}
+	}
+
+	// Seperate regular and tail calls.
+	for _, async := range c.asyncFuncs {
+		// Find all tail calls (of any kind).
+		tails := map[llvm.Value]struct{}{}
+		for ret, kind := range async.returns {
+			switch kind {
+			case returnVoidTail, returnTail, returnDeadTail, returnAlternateTail, returnDitchedTail, returnDelayedValue:
+				// This is a tail return. The previous instruction is a tail call.
+				tails[llvm.PrevInstruction(ret.LastInstruction())] = struct{}{}
+			}
+		}
+
+		// Find all regular calls.
+		regulars := map[llvm.Value]struct{}{}
+		for call := range async.calls {
+			if _, ok := tails[call]; ok {
+				// This is a tail call.
+				continue
+			}
+
+			regulars[call] = struct{}{}
+		}
+
+		async.tailCalls = tails
+		async.normalCalls = regulars
+	}
+}
+
+// lowerFuncsPass lowers all functions, turning them into coroutines if necessary.
+func (c *coroutineLoweringPass) lowerFuncsPass() {
+	for _, fn := range c.asyncFuncs {
+		if fn.fn == c.pause {
+			// Skip. It is an intrinsic.
+			continue
+		}
+
+		if len(fn.normalCalls) == 0 {
+			// No suspend points. Lower without turning it into a coroutine.
+			c.lowerFuncFast(fn)
+		} else {
+			// There are suspend points, so it is necessary to turn this into a coroutine.
+			c.lowerFuncCoro(fn)
+		}
+	}
+}
+
+func (async *asyncFunc) hasValueStoreReturn() bool {
+	for _, kind := range async.returns {
+		switch kind {
+		case returnNormal, returnAlternateTail, returnDelayedValue:
+			return true
+		}
+	}
+
+	return false
+}
+
+// heapAlloc creates a heap allocation large enough to hold the supplied type.
+// The allocation is returned as a raw i8* pointer.
+func (c *coroutineLoweringPass) heapAlloc(t llvm.Type, name string) llvm.Value {
+	sizeT := c.alloc.FirstParam().Type()
+	ptrT := llvm.PointerType(t, 0)
+	offsetPtr := llvm.ConstGEP(llvm.ConstNull(ptrT), []llvm.Value{llvm.ConstInt(sizeT, 1, false)})
+	size := llvm.ConstPtrToInt(offsetPtr, sizeT)
```

Comment:
You can simplify this using `TargetData.TypeAllocSize`. You can create a TargetData with `llvm.NewTargetData(c.mod.DataLayout())`